`podman generate spec` segfaults for pod without infra container

### Issue Description

When generating Specgen JSON, a segfault occurs [here](https://github.com/containers/podman/blob/v4.9.0/pkg/specgen/generate/pod_create.go#L371-L373) if the pod does not have an infra container.

Moving those lines inside the preceding `if pod.HasInfraContainer() {}` block resolves the segfault, but it's unclear if additional handling is needed for this particular case.

### Steps to reproduce the issue

```podman generate spec $(podman pod create --share=none)```

### Describe the results you received

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x298 pc=0x55f795d69d05]

goroutine 1 [running]:
panic({0x55f7965e4da0?, 0x55f7973998a0?})
	/usr/lib/golang/src/runtime/panic.go:1017 +0x3ac fp=0xc000559938 sp=0xc000559888 pc=0x55f794debd8c
runtime.panicmem(...)
	/usr/lib/golang/src/runtime/panic.go:261
runtime.sigpanic()
	/usr/lib/golang/src/runtime/signal_unix.go:861 +0x378 fp=0xc000559998 sp=0xc000559938 pc=0x55f794e03778
github.com/containers/podman/v4/pkg/specgen/generate.PodConfigToSpec(0xc00055e5a0, 0xc00051c380, 0xc0003c5800, {0x7ffd9569e2d3?, 0x0?})
	/builddir/build/BUILD/podman-4.9.0/pkg/specgen/generate/pod_create.go:371 +0x5c5 fp=0xc000559a70 sp=0xc000559998 pc=0x55f795d69d05
github.com/containers/podman/v4/pkg/domain/infra/abi.(*ContainerEngine).GenerateSpec(0xc00054c5a0, {0x0?, 0x0?}, 0xc0001fe780)
	/builddir/build/BUILD/podman-4.9.0/pkg/domain/infra/abi/generate.go:60 +0x14c fp=0xc000559b38 sp=0xc000559a70 pc=0x55f795dae94c
github.com/containers/podman/v4/cmd/podman/generate.spec(0x55f7973baf40?, {0xc00058bba0?, 0x1?, 0x1?})
	/builddir/build/BUILD/podman-4.9.0/cmd/podman/generate/spec.go:54 +0x9d fp=0xc000559ba8 sp=0xc000559b38 pc=0x55f795eb3ebd
github.com/spf13/cobra.(*Command).execute(0x55f7973baf40, {0xc0000400f0, 0x1, 0x1})
	/builddir/build/BUILD/podman-4.9.0/vendor/github.com/spf13/cobra/command.go:983 +0xabc fp=0xc000559d48 sp=0xc000559ba8 pc=0x55f795345ebc
github.com/spf13/cobra.(*Command).ExecuteC(0x55f7973b9560)
	/builddir/build/BUILD/podman-4.9.0/vendor/github.com/spf13/cobra/command.go:1115 +0x3ff fp=0xc000559e20 sp=0xc000559d48 pc=0x55f79534677f
github.com/spf13/cobra.(*Command).Execute(...)
	/builddir/build/BUILD/podman-4.9.0/vendor/github.com/spf13/cobra/command.go:1039
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/builddir/build/BUILD/podman-4.9.0/vendor/github.com/spf13/cobra/command.go:1032
main.Execute()
	/builddir/build/BUILD/podman-4.9.0/cmd/podman/root.go:115 +0xb8 fp=0xc000559ea8 sp=0xc000559e20 pc=0x55f795fe93f8
main.main()
	/builddir/build/BUILD/podman-4.9.0/cmd/podman/main.go:60 +0x467 fp=0xc000559f40 sp=0xc000559ea8 pc=0x55f795fe8b27
runtime.main()
	/usr/lib/golang/src/runtime/proc.go:267 +0x2d2 fp=0xc000559fe0 sp=0xc000559f40 pc=0x55f794deec52
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc000559fe8 sp=0xc000559fe0 pc=0x55f794e22f21
```

### Describe the results you expected

Specgen JSON output is returned

### podman info output

```yaml
podman-4.9.0-1.fc39.x86_64
```


### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

No

### Additional environment details

Additional environment details

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

--- a/pkg/specgen/generate/pod_create.go
+++ b/pkg/specgen/generate/pod_create.go
@@ -356,14 +356,15 @@ func PodConfigToSpec(rt *libpod.Runtime, spec *specgen.PodSpecGenerator, infraOp
 			return nil, err
 		}
 
+		if len(spec.InfraContainerSpec.Image) > 0 {
+			spec.InfraImage = spec.InfraContainerSpec.Image
+		}
+
 		spec.Name = name
 	}
 
 	// need to reset hostname, name etc of both pod and infra
 	spec.Hostname = ""
 
-	if len(spec.InfraContainerSpec.Image) > 0 {
-		spec.InfraImage = spec.InfraContainerSpec.Image
-	}
 	return pod, nil
 }

--- a/test/e2e/generate_spec_test.go
+++ b/test/e2e/generate_spec_test.go
@@ -57,4 +57,14 @@ var _ = Describe("Podman generate spec", func() {
 		session.WaitWithDefaultTimeout()
 		Expect(session).Should(ExitCleanly())
 	})
+
+	It("generate spec for pod with no infra container", func() {
+		session := podmanTest.Podman([]string{"pod", "create", "--share", "none", "--name", "podspecgen2"})
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(ExitCleanly())
+
+		session = podmanTest.Podman([]string{"generate", "spec", "--compact", "podspecgen2"})
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(ExitCleanly())
+	})
 })

